src/transport/smtp/client/connection.rs

```diff
@@ -82,9 +82,32 @@ impl SmtpConnection {
         // Mail
         let mut mail_options = vec![];
 
-        if self.server_info().supports_feature(Extension::EightBitMime) {
+        // Internationalization handling
+        //
+        // * 8BITMIME: https://tools.ietf.org/html/rfc6152
+        // * SMTPUTF8: https://tools.ietf.org/html/rfc653
+
+        // Check for non-ascii addresses and use the SMTPUTF8 option if any.
+        if envelope.has_non_ascii_addresses() {
+            if !self.server_info().supports_feature(Extension::SmtpUtfEight) {
+                // don't try to send non-ascii addresses (per RFC)
+                return Err(Error::Client(
+                    "Envelope contains non-ascii chars but server does not support SMTPUTF8",
+                ));
+            }
+            mail_options.push(MailParameter::SmtpUtfEight);
+        }
+
+        // Check for non-ascii content in message
+        if !email.is_ascii() {
```

Comment:
:thinking: The email should already be ascii, unless the user manually built the `Body` through [`Body::dangerous_pre_encoded`](https://docs.rs/lettre/0.10.0-alpha.5/lettre/message/struct.Body.html#method.dangerous_pre_encoded) or didn't go through the `lettre::message` builder. But I guess if we want to be as correct as possible with the implementation this extra check makes sense.

Reply:
indeed, in case the message is not built with our builder.